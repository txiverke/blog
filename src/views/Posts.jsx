// @flow

import React from 'react'
import Helmet from 'react-helmet'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import PostList from '../components/PostList'
import Loader from '../components/Loader'
import ShowMsg from '../components/ShowMsg'
import ButtonCorner from '../components/ButtonCorner'
import Search from '../containers/Search'
import { loadPostData, checkPostsTags } from '../actions/postActionCreators'
import config from '../config'
import { getSlug, isEqual } from '../utils/helpers'

class Posts extends React.Component {
  state = {
    tags: [],
    posts: [],
  }

  props: {
    dispatch: Function,
    posts: Data,
    tags: Tag,
  }

  componentDidMount() {
    const { dispatch, posts } = this.props

    if (posts.data.length === 0) {
      dispatch(loadPostData())
    }
  }

  componentWillReceiveProps(nextProps) {
    const { tags, dispatch, posts  } = nextProps
    if (!isEqual(tags.data, this.state.tags)) {
      this.setState({ tags: [...tags.data] })
      this.renderProperPosts(tags.data)
    }
  }

  getTags(data) {
    const listOfArr = data.map(item =>  item.tags.split(','))
    return Array.from(new Set([].concat(...listOfArr).map(item => item.trim())))
  }
  
  renderProperPosts(tags) {
    const { data } = this.props.posts
    const postsTagged = []
    
    tags.filter(tag => {
      return data.filter(item => {
        if (item.tags.includes(tag)) {
          postsTagged.push(item)
        }
      })
    })
      
    this.setState({ posts: Array.from(new Set(postsTagged)) })
  }

  render() {
    const { completed, data, message, error } = this.props.posts
    const { posts } = this.state

    if (completed) {
      const tags = this.getTags(data)

      return (
        <section className="app-article-wrapper">
          <Helmet 
            title="Posts" 
            meta={[
              { name:"description", content: "Posts about Frontend development" },
              { property: "og:title", content: "Posts about Frontend development" },
            ]}
          />
          {error && <ShowMsg message={message} error={error} next={completed} />}
          <ButtonCorner label="Back" />
          <Search tagsToRender={tags} />
          <PostList data={posts} />
        </section>
      )
    }

    return (
      <div className="app-view app-view-centered">
        <Loader msg={message} />
      </div>
    )
  }
} 

const mapStateToProps = state => ({ 
  posts: state.posts,
  tags: state.tags 
})
  
export default connect(mapStateToProps)(Posts)