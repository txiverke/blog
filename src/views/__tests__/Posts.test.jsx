import React from 'react'
import { Unwrapped as UnwrappedPosts } from '../Posts'
import { shallow, render } from 'enzyme'

describe('<Posts />', () => {
  it('should render itself', () => {
    const posts = {
      completed: true,
      data: [{ 
        created: "2018-02-16T00:25:46.636Z",
        _id: "5a86250a6e51f10ae32d5bc1",
        title: "Just a new post",
        content: "Just a new postJust a new postJust a new post",
        tags: "Just a new post",
        link: "https://google.com",
        background: "69372c7db575f3e1483e66bbe30d55b91518781651399.jpeg",
        __v: 0,
       }],
      message: 'string',
      error: false
    }
    const tags = ['']
    const language = { current: 'eng' }
    const component = shallow(<UnwrappedPosts posts={posts} tags={tags} language={language} />)
    expect(component).toMatchSnapshot()
  })

})