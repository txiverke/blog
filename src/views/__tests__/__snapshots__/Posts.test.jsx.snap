// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> should render itself 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Posts
    language={
        Object {
            "current": "eng",
          }
    }
    posts={
        Object {
            "completed": true,
            "data": Array [
              Object {
                "__v": 0,
                "_id": "5a86250a6e51f10ae32d5bc1",
                "background": "69372c7db575f3e1483e66bbe30d55b91518781651399.jpeg",
                "content": "Just a new postJust a new postJust a new post",
                "created": "2018-02-16T00:25:46.636Z",
                "link": "https://google.com",
                "tags": "Just a new post",
                "title": "Just a new post",
              },
            ],
            "error": false,
            "message": "string",
          }
    }
    tags={
        Array [
            "",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
          meta={
                    Array [
                              Object {
                                "content": "POSTS",
                                "name": "description",
                              },
                              Object {
                                "content": "POSTS",
                                "property": "og:title",
                              },
                            ]
          }
          title="POSTS"
/>,
        <ButtonBack
          label="Back"
          route="/"
/>,
        false,
        <Connect(SearchContainer)
          tagsToRender={
                    Array [
                              "Just a new post",
                            ]
          }
/>,
        <h1
          className="hidden"
>
          POSTS
</h1>,
        <PostList
          DIC={
                    Object {
                              "ABOUT-ME": "ABOUT ME",
                              "BACK": "Back",
                              "BACK_TO": "Back to the landing page",
                              "CONTACT": "CONTACT",
                              "DESCRIPTION": "Hi I'm Xavi, a front-end developer from Lleida (Catalunya) currently living with my wife and our dog and two cats in Ferndale (Michigan).",
                              "DOWNLOADED_RESUME": "Résumé downloaded",
                              "DOWNLOAD_RESUME": "Download Résumé",
                              "EDIT": "EDIT",
                              "FRONTEND": "Front-end developer",
                              "NOT_EXIST": " does not exist.",
                              "NO_POSTS_MSG": "Press a tag button to see content.",
                              "PAGE_NOT_FOUND": "Page not found",
                              "POSTED_AT": "Posted at ",
                              "POSTS": "POSTS",
                              "PROJECTS": "PROJECTS",
                              "RESUME": "Thanks for downloading my Résumé.",
                              "SORRY": "Sorry, but the url: ",
                              "WELCOME": "Welcome to ",
                            }
          }
          list={Array []}
/>,
      ],
      "className": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "defer": true,
          "encodeSpecialCharacters": true,
          "meta": Array [
            Object {
              "content": "POSTS",
              "name": "description",
            },
            Object {
              "content": "POSTS",
              "property": "og:title",
            },
          ],
          "title": "POSTS",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "label": "Back",
          "route": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "tagsToRender": Array [
            "Just a new post",
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "POSTS",
          "className": "hidden",
        },
        "ref": null,
        "rendered": "POSTS",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "DIC": Object {
            "ABOUT-ME": "ABOUT ME",
            "BACK": "Back",
            "BACK_TO": "Back to the landing page",
            "CONTACT": "CONTACT",
            "DESCRIPTION": "Hi I'm Xavi, a front-end developer from Lleida (Catalunya) currently living with my wife and our dog and two cats in Ferndale (Michigan).",
            "DOWNLOADED_RESUME": "Résumé downloaded",
            "DOWNLOAD_RESUME": "Download Résumé",
            "EDIT": "EDIT",
            "FRONTEND": "Front-end developer",
            "NOT_EXIST": " does not exist.",
            "NO_POSTS_MSG": "Press a tag button to see content.",
            "PAGE_NOT_FOUND": "Page not found",
            "POSTED_AT": "Posted at ",
            "POSTS": "POSTS",
            "PROJECTS": "PROJECTS",
            "RESUME": "Thanks for downloading my Résumé.",
            "SORRY": "Sorry, but the url: ",
            "WELCOME": "Welcome to ",
          },
          "list": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
            meta={
                        Array [
                                    Object {
                                      "content": "POSTS",
                                      "name": "description",
                                    },
                                    Object {
                                      "content": "POSTS",
                                      "property": "og:title",
                                    },
                                  ]
            }
            title="POSTS"
/>,
          <ButtonBack
            label="Back"
            route="/"
/>,
          false,
          <Connect(SearchContainer)
            tagsToRender={
                        Array [
                                    "Just a new post",
                                  ]
            }
/>,
          <h1
            className="hidden"
>
            POSTS
</h1>,
          <PostList
            DIC={
                        Object {
                                    "ABOUT-ME": "ABOUT ME",
                                    "BACK": "Back",
                                    "BACK_TO": "Back to the landing page",
                                    "CONTACT": "CONTACT",
                                    "DESCRIPTION": "Hi I'm Xavi, a front-end developer from Lleida (Catalunya) currently living with my wife and our dog and two cats in Ferndale (Michigan).",
                                    "DOWNLOADED_RESUME": "Résumé downloaded",
                                    "DOWNLOAD_RESUME": "Download Résumé",
                                    "EDIT": "EDIT",
                                    "FRONTEND": "Front-end developer",
                                    "NOT_EXIST": " does not exist.",
                                    "NO_POSTS_MSG": "Press a tag button to see content.",
                                    "PAGE_NOT_FOUND": "Page not found",
                                    "POSTED_AT": "Posted at ",
                                    "POSTS": "POSTS",
                                    "PROJECTS": "PROJECTS",
                                    "RESUME": "Thanks for downloading my Résumé.",
                                    "SORRY": "Sorry, but the url: ",
                                    "WELCOME": "Welcome to ",
                                  }
            }
            list={Array []}
/>,
        ],
        "className": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defer": true,
            "encodeSpecialCharacters": true,
            "meta": Array [
              Object {
                "content": "POSTS",
                "name": "description",
              },
              Object {
                "content": "POSTS",
                "property": "og:title",
              },
            ],
            "title": "POSTS",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "label": "Back",
            "route": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "tagsToRender": Array [
              "Just a new post",
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "POSTS",
            "className": "hidden",
          },
          "ref": null,
          "rendered": "POSTS",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "DIC": Object {
              "ABOUT-ME": "ABOUT ME",
              "BACK": "Back",
              "BACK_TO": "Back to the landing page",
              "CONTACT": "CONTACT",
              "DESCRIPTION": "Hi I'm Xavi, a front-end developer from Lleida (Catalunya) currently living with my wife and our dog and two cats in Ferndale (Michigan).",
              "DOWNLOADED_RESUME": "Résumé downloaded",
              "DOWNLOAD_RESUME": "Download Résumé",
              "EDIT": "EDIT",
              "FRONTEND": "Front-end developer",
              "NOT_EXIST": " does not exist.",
              "NO_POSTS_MSG": "Press a tag button to see content.",
              "PAGE_NOT_FOUND": "Page not found",
              "POSTED_AT": "Posted at ",
              "POSTS": "POSTS",
              "PROJECTS": "PROJECTS",
              "RESUME": "Thanks for downloading my Résumé.",
              "SORRY": "Sorry, but the url: ",
              "WELCOME": "Welcome to ",
            },
            "list": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
